### DISCUSSION TITLE
Renovate removes pip package suffixes

### DISCUSSION THREAD

**Discussion by davidspek** [renovatebot/renovate#25057] [Suggest an Idea]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

5.0.1

### If you're self-hosting Renovate, select which platform you are using.

github.com

### Was this something which used to work for you, and then stopped?

I am trying to get this working for the first time

### Describe the problem

Although uncommon, some pip packages have suffixes. One example of this is PyTorch where there are different version suffixes for CPU (`+cpu`) and CUDA (`+cu118`) installation, for CPU and GPU accelerated environments respectively. When Renovate tries to update the PyTorch dependencies, it is removing the suffixes from the versions. The relevant PR in question is https://github.com/pluralsh/kubeflow-notebooks/pull/298.

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

Placeholder value, please select the correct response from the dropdown

**Comment by rarkins** [comment]
Is there a canonical description of how such suffixes work in Python/Pep440? Or is this just a convention some users/packages adopted?

I _think_ they're referred to as "local version identifiers" and not originally intended as compatibility indicators: https://peps.python.org/pep-0440/#local-version-identifiers

I'm ok if users have run off and defined their own convention, as we always do our best to support. But I'm trying to work out if this is really a problem/bug or not.

**Reply by davidspek** [reply]
From what I know those are indeed "local version identifiers". I'd expect renovate to match/keep the suffixes similar to what it does for docker (like the common `-alpine3.17` suffixes) when it tries to find new versions. But I'm not sure how Renovate should handle this in the most correct way possible.

**Reply by rarkins** [reply]
Your expectation is that Renovate takes a non standard approach to python versioning _by default_? And people using it properly need to add to config to restore standardized behavior?

**Reply by davidspek** [reply]
Not exactly. I'd expect it to keep suffixes of versions, but like I mentioned I'm not sure how that would be done in a correct way that doesn't mess with the standard versioning logic.

**Reply by rarkins** [reply]
I've recategorized this from "Report a problem" to "Request a feature".

If we want to be strict, we'd say that everyone doing this is misusing local identifiers and by default we behave correctly. If we want to be flexible then we'd try to find a way to identify only the type of string which is likely a compatibility string and not likely to be a properly used local identifier. An all-character one like `+cpu` might be good for that, but something like `+cu18` is a bit harder to determine. Can we realistically guess these accurately enough not to inconvenience users of packages using local identifiers for the intended purpose?

BTW you should be able to fix this with config already using the `versionCompatibility` feature - there are some examples in the docs you can follow

**Reply by davidspek** [reply]
It actually looks like PyTorch has changed their packaging and installation instructions so this issue isn't hit anymore. Given that they were likely misusing local identifiers and probably not many other packages are doing this I think this can likely be closed or ignored for now.
